Fix: apply clang-tidy fixes

caic99 opened the pull request:
### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #3915

### Unit Tests and/or Case Tests for my changes
- A unit test which is expected to exit with assertion error has been fixed.

### What's changed?
This PR applies clang-tidy rules on existing codes to improve code quality. I've carefully reviewed the modifications.

caic99 committed:
update clang-tidy config for fixing

caic99 committed:
apply `run-clang-tidy-12 -fix` and adjust some false cases

caic99 committed:
adjust UT

caic99 committed:
fix unitcell ut

caic99 committed:
fix read atoms

caic99 committed:
patch paw

caic99 committed:
fix paw

caic99 committed:
fix paw init

caic99 committed:
fix read pp UT

haozhihan commented:
According to [Commit message guidelines](https://abacus.deepmodeling.com/en/latest/CONTRIBUTING.html#commit-message-guidelines), I think using keyword `Style` might be more appropriate.

caic99 commented:
> According to [Commit message guidelines](https://abacus.deepmodeling.com/en/latest/CONTRIBUTING.html#commit-message-guidelines), I think using keyword `Style` might be more appropriate.

This PR makes actual changes to the code.

kirk0830 commented on source/module_io/binstream.cpp:
```diff
@@ -37,18 +37,12 @@ void Binstream::open(const std::string filename,const char *op)
-		return true;
-	else
-		return false;
+	return fileptr==NULL;
```
I  am a little bit confused here, why use `nullptr` and `NULL` in the same file? Seems more reasonable choice would be using `nulltpr` for all?

caic99 replied:
> I am a little bit confused here, why use `nullptr` and `NULL` in the same file? Seems more reasonable choice would be using `nulltpr` for all?

Only one fix can be applied at once.

```suggestion
	return fileptr==nullptr;
```

kirk0830 commented on source/module_io/cal_dos.cpp:
```diff
@@ -41,9 +41,9 @@ bool ModuleIO::calculate_dos
+	if(emax_ev < emin_ev)
 	{
 		ModuleBase::WARNING("ModuleIO::calculate_dos","emax_ev < emin_ev");
 		return 0;
```
should it be `return false;` ?

caic99 replied:
> should it be `return false;` ?

```suggestion
		return false;
```

kirk0830 commented on source/module_io/cal_dos.h:
```diff
@@ -6,8 +6,8 @@
 	bool calculate_dos(const int &is,		
-		const std::string &fn,// file address for DOS.
-		const std::string &fn1,// file address for DOS_smearing.
+		const std::string &fa,// file address for DOS.
```
this change results from wrong annotation, "file address", instead it seems to be more appropriate if as `file name` or something

caic99 replied:
to match the definition

kirk0830 commented on source/module_io/istate_envelope.h:
```diff
@@ -59,7 +59,7 @@ class IState_Envelope
                Gint_k& gk,
                int& out_wfc_pw,
-               int& out_wfc_r,
+               int& out_wf_r,
```
why???

caic99 replied:
To match the name in definition

jinzx10 commented on source/module_base/intarray.h:
I think the following merely default-initializes bound6; bound1-5 are still uninitialized.
```
int bound1, bound2, bound3, bound4, bound5, bound6{};
```

caic99 committed:
fix gpu branch

mohanchen commented on source/module_base/abfs-vector3_order.cpp:
```diff
@@ -27,10 +27,10 @@ template<>
 {
 	if      ( v1.x < v2.x ) return true;
-	else if ( v1.x > v2.x ) return false;
+	if ( v1.x > v2.x ) return false;
```
I feel the logic has been changed here.

caic99 replied:
There's no need to use `else` after `return`, see https://clang.llvm.org/extra/clang-tidy/checks/readability/else-after-return.html

mohanchen commented on source/module_base/complexarray.h:
```diff
@@ -117,7 +117,7 @@ void scaled_sum(std::complex <double> c1, const ComplexArray &cd1,
 
 /// out[i] = a1[i] * in2[i]
-void point_mult(ComplexArray &a1, ComplexArray &in2, ComplexArray &out);
+void point_mult(ComplexArray &in1, ComplexArray &in2, ComplexArray &out);
```
I feel the meaning has been changed here

caic99 replied:
The name is changed in consistent with that in `complexarray.cpp`.

mohanchen commented on source/module_base/inverse_matrix.h:
```diff
@@ -20,12 +20,12 @@ class Inverse_Matrix_Complex
+	double *e{};
+	int lwork{};
+	std::complex<double> *work2{};
+	double* rwork{};
```
why not use nullptr here?

caic99 replied:
Using `{}` is identical, as it [initializes the pointer to nullptr](https://en.cppreference.com/w/cpp/language/zero_initialization#:~:text=A%20zero%2Dinitialized%20pointer%20is%20the%20null%20pointer%20value%20of%20its%20type%2C%20even%20if%20the%20value%20of%20the%20null%20pointer%20is%20not%20integral%20zero.).

mohanchen replied:
I am not looking identical ways of writing C++ codes, I just consider this new way of initializing variables make reading more difficult for some people, we should have a unified way. For example, double *rwork = nullptr;

mohanchen commented on source/module_base/kernels/math_op.cpp:
```diff
@@ -1,11 +1,14 @@
 template <typename FPTYPE>
 __inline__
-FPTYPE __fact(const int n) {
+FPTYPE _fact(const int n) {
```
is there any special reason to change the function name from __ to _

caic99 replied:
The C++ standard reserves names starting with __.

mohanchen commented on source/module_base/math_chebyshev.h:
```diff
@@ -187,7 +187,7 @@ class Chebyshev
 public:
 	//Members:
-    int norder;   // order of Chebyshev expansion
+    int norder{};   // order of Chebyshev expansion
```
why not initialize it will 0?

caic99 replied:
It will, please see https://en.cppreference.com/w/cpp/language/zero_initialization

mohanchen replied:
I mean, why not write the code "int norder = 0" ?

caic99 replied:
> I mean, why not write the code "int norder = 0" ?

Using bracelet is the only supported grammar for [this fix](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-default-member-init.html) initializing members in declaration. There is [another fix](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/init-variables.html) available in your expected grammar, but the modifications will be much more.

mohanchen commented on source/module_base/math_sphbes.cpp:
```diff
@@ -258,7 +258,7 @@ void Sphbes::BESSJY(double x, double xnu, double *rj, double *ry, double *rjp, d
 {
     if (a > b) return a;
-    else return b;
+    return b;
```
logic changes here

mohanchen commented on source/module_base/module_container/ATen/ops/linalg_op.h:
```diff
@@ -137,7 +137,7 @@ struct inflate_op {
     void operator()(
         const Tensor& input,
-        const std::vector<int64_t>& stride,
+        const std::vector<int64_t>& inflate,
```
what's the reason to change the variable's name?

haozhihan replied:
This is because the corresponding variable name in the implementation of the function in the `.cpp` file is `inflate`. So, it is to keep the declaration and implementation of the function consistent. 

LGTM.

mohanchen commented on source/module_base/opt_DCsrch.cpp:
```diff
@@ -171,18 +171,18 @@ int dcsrch(double& stp,
-                                        bool&,
-                                        double&,
-                                        double&);
+    extern /* Subroutine */ void dcstep(double& /*stx*/,
```
looks strange here

caic99 replied:
Add variable names according to declaration to improve code readability.

mohanchen replied:
We don't recommend writing "/* */"

mohanchen commented on source/module_base/parallel_global.cpp:
```diff
@@ -285,52 +281,18 @@ void Parallel_Global::finalize_mpi()
 // CALL Function : divide_pools
 //----------------------------------------------------------
     Parallel_Global::divide_pools();
-
```
why delete these sentences?

caic99 replied:
will add them back

mohanchen commented on source/module_base/spherical_bessel_transformer.cpp:
```diff
@@ -34,39 +34,39 @@ class SphericalBesselTransformer::Impl
-                double* const out,
-                const int p = 0,
-                const bool deriv = false
+    void direct(int l,
```
why delete const?

caic99 replied:
Parameter 'l' is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions. See https://clang.llvm.org/extra/clang-tidy/checks/readability/avoid-const-params-in-decls.html

mohanchen replied:
Will it be good to be consistent between declaration and definitions? I don't see a reason to make these two codes different

caic99 replied:
> Will it be good to be consistent between declaration and definitions? I don't see a reason to make these two codes different

I see. The point is that declarations should be put in header files; the declarations here are not top-level, hence `const` qualifier has no effect here, making a possible definitions without const possible. @jinzx10 What's your consideration here?

However, in this context, I agree with you that preserve `const` would be better.

mohanchen reviewed (commented):
I have looked through dozens of files, most of them look good to me. Howver, I don't understand some modifications, as can be seen with my comments.

mohanchen committed:
Merge branch 'develop' into clang-tidy-fix

jinzx10 commented:
I went through the files in module_base, module_basis and module_cell. Apart from the mistaken removal of some gpu code in pw_basis_sup.cpp (which has been fixed in a commit), clang-tidy indeed tidies up a lot of code. Here's a brief list of what I record:

1. add const& to function argument that are not altered in the body
This can save many uncessary copy and is definitely a good practice. 

2. use std::move when copying variables that are no longer needed
This can also prevent unnecessary copy (given that the move constructor / move assignment operator is available)

3. delete "return" for void functions
4. delete "void" in function parameter list
These are basically a matter of style. Maybe we should choose one and stick with it. (PS: In C, a function declared as func() can take any arguments, while a function that takes no argument should be declared as func(void). In C++, func() is equivalent to func(void) so in some sense func(void) is more compatible and func() is more C++)

5. reserve a std::vector before looping over push_back
This is definitely a good practice if one can somehow know or estimate the final size in advance. If the objects are expensive to copy, or if there're a lots of elements to push back, reserving a vector could save a lot of time.

6. replacing NULL or 0 with nullptr
good practice by c++ standard

7. more concise code when using bool variables
If x is a bool, if statement can simply be if(x) instead of if (x == true) (or if(!x) instead of if (x == false)). I believe these two styles have no essential difference as the compiler would be smart enough to generate the same code, but I personally do prefer the shorter one. Another case is
```
if (some_condition)
    x = true;
else
    x = false;
```
which can be simply
```
x = some_condition;
```

8. use the member function empty() when checking if a stl container is empty
Technically empty() provided by the container is the fastest way to check if a container is empty. Take vector as an example, x.size() == 0 actually involves two steps: (1) subtracting "begin" from "finish" and (2) compare the result with 0, while empty() simply performs one step "begin == finish". Though the difference is marginal and might be optimized by the compiler, I think it's always better to stick with the better one.

9. make member functions static whenever possible
This is important for unit test. Static member function can be called (thus tested) without constructing object.

10. initialize POD type variables (int, double, pointer, ...)
This is very important to avoid many unexpected behaviors. An uninitialized int/double/pointer could be any random value. What's even worse is that the piece of memory allocated to those uninitialized variables are used before and is occupied by some meaingful value. In that case, even if the code has some logical error, such uninitialized variables might accidentally make things work. There's actually a fresh example in this PR:  in source/module_cell/test/read_pp_test.cpp, HeaderErr2015, an extra "EXPECT_DEATH" has to be added to the tested function which is supposed to fail. The reason that the function could normally terminate before this PR is that the test involves an assertion on an uninitialized variable which was set in the last unit test HeaderErr2013. Even though test objects are intialized for each test in unit testing, uninitialized variables could actually "inherit" values from previous tests if memory are allocated the same way.

11. no need to initialize a string with "" (and other stl containers to explicit empty)
As opposed to POD types, default intializations of stl containers are all properly handled - e.g., std::string is default initialized to "" -  so there's no need to specify.

12. when finding a single character in a string, make sure to pass a char instead of a string with one character:
```
str.find(':'); // good
str.find(":"); // not as good, though practically the difference might be neglible
```

13. in an if-else statement, removing the "else-branch" if the "if-branch" contains something that terminates the code flow:
for example, clang-tidy constantly replaces
```
if (condition)
    return some_variable;
else
    do_something_else();
```
with
```
if (condition)
    return some_variable;
do_something_else();
```
I think this is another matter of style which I have no particular preference. However, if both branches return something, I personally prefer the ternary operator:
```
return (condition) ? A : B;
```
For the three modules I was assigned, I think the changes made by this PR looks good (with some style issue that may or may not need discussion). The only real issue is that this PR messes up quite a few indentation, which might be fixed by another PR of clang-format, I guess?

kirk0830 commented on source/module_hamilt_pw/hamilt_pwdft/operator_pw/meta_pw.h:
```diff
@@ -26,7 +26,7 @@ class Meta<OperatorPW<T, Device>> : public OperatorPW<T, Device>
     using Real = typename GetTypeReal<T>::type;
     public:
-      Meta(Real tpiba2_in,
+      Meta(Real tpiba_in,
```
I am not sure but it seems `tpiba` is not equivalent with `tpiba2`, the latter seems to be square of the former.

mohanchen replied:
I agree

kirk0830 commented on source/module_hamilt_pw/hamilt_pwdft/operator_pw/velocity_pw.h:
```diff
@@ -35,8 +35,8 @@ class Velocity
              const int n_npwx,
-             const std::complex<double>* tmpsi_in,
-             std::complex<double>* tmhpsi,
+             const std::complex<double>* psi0,
```
this seems not a good idea...

kirk0830 commented on source/module_hamilt_pw/hamilt_pwdft/stress_func_cc.cpp:
firstly I am fully puzzled about "what is 'cc' the contribution of stress", finally I realize it is `nlcc` instead of `cc`... This filename is at least partially unacceptable. @mohanchen

caic99 commented:
> The only real issue is that this PR messes up quite a few indentation, which might be fixed by another PR of clang-format, I guess?

Running `clang-format` will modify codes significantly, so this PR does not execute it.

caic99 committed:
fix tests in parallel_global.cpp

caic99 committed:
not applying `readability-else-after-return`

kirk0830 reviewed (commented):
I have reviewed files in module_hamilt_pw, module_io, module_psi.
There are mainly several kinds of code polishment I have observed:
1. change "pass by value" to "pass by ref" or "pass by value then swap".
2. add `static` keyword ahead of function that does not rely on member variables.
3. delete useless `return` in void function.
4. change some `0` and `1` to `false` or `true`, correspondingly, avoid implicit type conversion
5. use list initialization on PART of member variables.
6. remove the use of operator `?:` in some boolean assignment cases.
7. add annotation behind namespace.
8. make nomenclature of varialbles consistent.
9. delete redundant `if not nullptr` sentence before deallocating one pointer
9. change the default initialization value `NULL` for pointer to `nullptr`

However, 
1. it (clang-tidy) indeed removed some annotation by mistake especially at the end of functions.
2. the indent seems to be not well-treated, but I guess it is because the mixed use of Tab and whitespaces
3. not fully change the if-else logic, like in psi_initializer_atomic.cpp:
```c++
std::complex<double> phase_factor(double arg, int mode)
{
    if(mode == 1) return std::complex<double>(cos(arg),0);
    if (mode == -1) return std::complex<double>(0, sin(arg));
    else if (mode == 0) return std::complex<double>(cos(arg), sin(arg));
// can also be `if` instead of `else if`
    else return std::complex<double>(1,0);
// can directly `return`
}
```

caic99 committed:
Update source/module_io/binstream.cpp

caic99 committed:
Update source/module_io/cal_dos.cpp

kirk0830 committed:
Merge branch 'develop' into clang-tidy-fix

caic99 commented:
Hi @mohanchen @kirk0830 ,
I've replied the conversations and modified codes w.r.t. our discussion. Would you please kindly review the PR? Feel free to comment if you have further questions. Many thanks.

mohanchen commented:
> Hi @mohanchen @kirk0830 , I've replied the conversations and modified codes w.r.t. our discussion. Would you please kindly review the PR? Feel free to comment if you have further questions. Many thanks.

Hi Cun, though most of the updates look good to me. However, I am deeply concerned about some modifications that change the parameter names, and the amount of modifications of codes is huge (522 files!). In addition, although I ask some questions, I don't get the answers I want (not the explanations for C++ grammar).

caic99 commented:
> Hi Cun, though most of the updates look good to me. However, I am deeply concerned about some modifications that change the parameter names, and the amount of modifications of codes is huge (522 files!). In addition, although I ask some questions, I don't get the answers I want (not the explanations for C++ grammar).

Hi @mohanchen ,
I fully understand your concern and am agree with you. 
This PR acts as an experiment to provide a possible solution for using automatic tools to improve code quality, and actually I'm not expecting it will be merged into trunk. However, I think the most important point here is that providing us a chance to discuss the rules in detail to maintain code quality.

mohanchen commented:
> > Hi Cun, though most of the updates look good to me. However, I am deeply concerned about some modifications that change the parameter names, and the amount of modifications of codes is huge (522 files!). In addition, although I ask some questions, I don't get the answers I want (not the explanations for C++ grammar).
> 
> Hi @mohanchen , I fully understand your concern and am agree with you. This PR acts as an experiment to provide a possible solution for using automatic tools to improve code quality, and actually I'm not expecting it will be merged into trunk. However, I think the most important point here is that providing us a chance to discuss the rules in detail to maintain code quality.

Thanks, I understand your thoughts better now. How about we close this PR but we label it as 'Useful Information', then we can use this as a reference for further discussion of how to make changes as we expect. @caic99

caic99 commented:
Hi all,
I've updated the clang-tidy rules discussed above.

> in an if-else statement, removing the "else-branch" if the "if-branch" contains something that terminates the code flow:
for example, clang-tidy constantly replaces

I've removed this rule removing else after a return statement. This is indeed optional, and will making incomplete changes.

> the indent seems to be not well-treated

This is the job of clang-format, and we may discuss it with another issue.

---

However, I still insist that some fixes are necessary, which we've also discussed above (and I kept them unresolved to show more details):
1. Consistency between declaration and definition
2. Zero-initialize variables

By consolidating a rule and applying them on each PR, the code quality is able to be assured and improved during refactor.
